turbo/jsonrpc/bor_snapshot.go

```diff
@@ -250,6 +257,79 @@ func (api *BorImpl) GetVoteOnHash(ctx context.Context, starBlockNr uint64, endBl
 	return true, nil
 }
 
+// SendRawTransactionConditional will add the signed transaction to the transaction pool.
+// The sender/bundler is responsible for signing the transaction
+func (api *BorImpl) SendRawTransactionConditional(ctx context.Context, encodedTx hexutility.Bytes, options types2.TransactionConditions) (common.Hash, error) {
+	txn, err := types.DecodeWrappedTransaction(encodedTx)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	// this has been moved to prior to adding of transactions to capture the
+	// pre state of the db - which is used for logging in the messages below
+	tx, err := api.db.BeginRo(ctx)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	defer tx.Rollback()
+
+	currentHeader := rawdb.ReadCurrentHeader(tx)
+	// Ensure we have an actually valid block
+	if currentHeader == nil {
+		return common.Hash{}, errUnknownBlock
+	}
+
+	tempBlockNumber := rpc.BlockNumber(currentHeader.Number.Int64())
+	tempBlockorHash := rpc.BlockNumberOrHash{BlockNumber: &tempBlockNumber}
+
+	readerTemp, err := rpchelper.CreateStateReader(ctx, tx, tempBlockorHash, 0, api.filters, api.stateCache, api.historyV3(tx), "")
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	currentState := state.New(readerTemp)
+
+	if err := stagedsync.ValidateTransactionConditions(&options, currentHeader, currentState, true); err != nil {
+		return common.Hash{}, err
+	}
+
+	// put options data in Tx, to use it later while block building
+	txn.PutOptions(&options)
```

Comment:
it seems to us that this property will not propagate to the mining stage

the new API call adds a transaction to the pool using the txpool GRPC `Add()` API:
[txPoolGrpcServer](https://github.com/ledgerwatch/interfaces/blob/master/txpool/txpool.proto#L90)

it is then added to the pool in the mining process here with `txPool.AddLocalTxs()`:
https://github.com/ledgerwatch/erigon/blob/devel/erigon-lib/txpool/txpool_grpc_server.go#L216

the mining process could be a separate process from the JSON RPC API server process

later on the mining stage grabs transactions from the pool using `YieldBest()` here:
https://github.com/ledgerwatch/erigon/blob/devel/eth/stagedsync/stage_mining_exec.go#L202

if we want to pass TransactionConditions from the API following the same route, we'll need to extend all 3 call points (GRPC Add, AddLocalTxs, YieldBest) with extra parameters, and add a new property to `TxSlot` or `metaTx` to hold it.

@AskAlexSharov , @mh0lt , is this analysis correct? is it a reasonable thing to do, or do we want a different approach for bor-specific transaction metadata?

Reply:
@battlmonstr you described well. Is this metadata encoded into txn rlp? Or it’s some side-data?

Reply:
If external: 
- txpool already has MetaTx wrapper for external data (or aggregates).
- Grpc methods: need extend. 

But: txpool itself does sorting txs from best to worst - by method Less. Maybe method Less must his new fields? (It knows current block number). Then miner will have less logic.

Anyway it feels useful feature that: miner can YeldBest with conditions - it may allow build more flexible miners. 


FYI: in our architecture view - miner was inside txpool process (always) - this is reason why there is no grpc between txpool and miner. 


About conditions type: plz check if possible use uint256 instead of big.Int.

Reply:
@AskAlexSharov Thanks for the suggestions. This metadata comes as a separate parameter of this new JSON RPC call, so not inside a txn encoding.

Reply:
Hey @battlmonstr - thanks for pointing this out. I'll update this so that the `options` are available to the miner.